contracts/examples/DXswapRelayer.sol

```diff
@@ -0,0 +1,387 @@
+        (uint reserveA, uint reserveB,) = IDXswapPair(dxSwapPair).getReserves();
+        uint256 windowTime = _consultOracleParameters(amountA, amountB, reserveA, reserveB, maxWindowTime);
+        orders[orderIndex].oracleId = oracleCreator.createOracle(windowTime, pair);
+        emit NewOrder(orderIndex, REMOVAL);
+    }
+
+    function executeOrder(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(orderIndex <= orderCount, 'DXswapRelayer: INVALID_ORDER');
+        require(!order.executed, 'DXswapRelayer: ORDER_EXECUTED');
+        require(oracleCreator.isOracleFinalized(order.oracleId) , 'DXswapRelayer: OBSERVATION_RUNNING');
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+
+        address tokenA = order.tokenA;
+        address tokenB = order.tokenB;
+        uint256 amountB;
+        amountB = oracleCreator.consult(
+          order.oracleId,
+          tokenA == address(0) ? IDXswapRouter(dxSwapRouter).WETH() : tokenA,
+          order.amountA 
+        );
+        uint256 amountA = oracleCreator.consult(order.oracleId, tokenB, order.amountB);
+        
+        /* Maximize token inputs */ 
+        if(amountA <= order.amountA){
+            amountA = amountA;
+            amountB = order.amountB;
+        } else {
+            amountB = amountB;
+            amountA = order.amountA;
+        }
+        uint256 minA = amountA.sub(amountA.mul(order.priceTolerance) / PARTS_PER_MILLION);
+        uint256 minB = amountB.sub(amountB.mul(order.priceTolerance) / PARTS_PER_MILLION);
+
+        order.executed = true;
+        if(order.action == PROVISION){
+            _pool(tokenA, tokenB, amountA, amountB, minA, minB);
+        } else if (order.action == REMOVAL){
+            address pair = _pair(tokenA, tokenB, dxSwapFactory);
+            _unpool(
+              tokenA, 
+              tokenB, 
+              pair, 
+              order.liquidity,
+              minA,
+              minB
+            );
+        }
+        emit ExecutedOrder(orderIndex);
+    }
+
+    // Updates a price oracle and sends a bounty to msg.sender
+    function updateOracle(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+        require(!oracleCreator.isOracleFinalized(order.oracleId) , 'DXswapRelayer: OBSERVATION_ENDED');
+        uint256 amountBounty = GAS_ORACLE_UPDATE.mul(tx.gasprice).add(BOUNTY);
+        require(address(this).balance >= amountBounty, 'DXswapRelayer: INSUFFICIENT_BALANCE');
+        (uint reserveA, uint reserveB,) = IDXswapPair(order.oraclePair).getReserves();
+        require(
+            reserveA >= order.minReserveA && reserveB >= order.minReserveB,
+            'DXswapRelayer: RESERVE_TO_LOW'
+        );
+        oracleCreator.update(order.oracleId);
+        TransferHelper.safeTransferETH(msg.sender, amountBounty);
```

Comment:
Like I mentioned above, what if we only refund if the balance is enough? If not lets skip the refund